### PR TITLE
refactor eds functions

### PR DESCRIPTION
**Pull request by ramaraochavali**
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/serviceregistry/serviceentry/servicediscovery.go [hunk 1]
```diff
@@ -316,19 +316,13 @@ func (s *ServiceEntryStore) serviceEntryHandler(_, curr config.Config, event mod
...
 
-	// If service entry is deleted, cleanup endpoint shards for services.
-	for _, svc := range deletedSvcs {
-		s.XdsUpdater.SvcUpdate(shard, string(svc.Hostname), svc.Attributes.Namespace, model.EventDelete)
+	changedServices := make([]*model.Service, 0, len(addedSvcs)+len(updatedSvcs)+len(deletedSvcs))
+	changedServices = append(changedServices, addedSvcs...)
+	changedServices = append(changedServices, deletedSvcs...)
+	changedServices = append(changedServices, updatedSvcs...)
+	for _, svc := range changedServices {
+		s.XdsUpdater.SvcUpdate(shard, string(svc.Hostname), svc.Attributes.Namespace, model.EventAdd)
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:325]: now we send EventAdd for deletes? Is that correct?

Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:321]: No. It is incorrect. Great catch. Reverted it.

File: pilot/pkg/serviceregistry/serviceentry/servicediscovery.go [hunk 2]
```diff
@@ -360,7 +361,7 @@ func (s *ServiceEntryStore) serviceEntryHandler(_, curr config.Config, event mod
 	fullPush := len(configsUpdated) > 0
 	// if not full push needed, at least one service unchanged
 	if !fullPush {
-		s.edsUpdate(serviceInstances, true)
+		s.queueEdsPush(serviceInstances)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:364]: could we call `edsUpdate` and `edsCacheUpdate`? 

queueEdsPush can not express this is a doing cache update -> push in searial

File: pilot/pkg/serviceregistry/serviceentry/servicediscovery.go [hunk 3]
```diff
@@ -600,9 +613,41 @@ func (s *ServiceEntryStore) edsUpdateInSerial(keys map[instancesKey]struct{}, pu
...
+		}
+	} else {
+		for k, eps := range endpoints {
+			s.XdsUpdater.EDSUpdate(shard, string(k.hostname), k.namespace, eps)
+		}
+	}
+}
+
+// buildEndpoints builds endpoints for the instance keys.
+func (s *ServiceEntryStore) buildEndpoints(keys map[instancesKey]struct{}) map[instancesKey][]*model.IstioEndpoint {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:649]: I am all for this abstraction

File: pilot/pkg/serviceregistry/serviceentry/servicediscovery.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:364]: Sorry. I did not understand. We already have edsUpate and edsCacheUpdate that call xds update functions. What are you suggesting here?

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:364]: I mean keep the name `edsUpdate` and `edsCacheUpdate`

Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/serviceentry/servicediscovery.go]: Changed a bit. PTAL

**Review by hzxuzhonghu — APPROVED** [review]
Much better

**Pull request merged**